packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -824,9 +825,46 @@ class _TextSelectionHandleOverlayState
         break;
     }
 
+    // On some platforms we may want to calculate the start and end handles
+    // separately so they scale for the selected content.
+    //
+    // For the start handle we compute the rectangles that encompass the range
+    // of the first full selected grapheme cluster at the beginning of the selection.
+    //
+    // For the end handle we compute the rectangles that encompass the range
+    // of the last full selected grapheme cluster at the end of the selection.
+    final InlineSpan span = widget.renderObject.text!;
+    final String text = span.toPlainText();
+    final int firstSelectedGraphemeExtent;
+    final int lastSelectedGraphemeExtent;
+    if (text.isEmpty || widget.renderObject.selection == null || !widget.renderObject.selection!.isValid || widget.renderObject.selection!.isCollapsed) {
+      // The call to selectedGraphemes.characters.first/last will throw a state
+      // error if the given string is empty, so fall back to first/last character
+      // range in this case.
+      //
+      // The call to widget.selection.textInside(text) will return a RangeError
+      // for a collapsed selection, fall back to this case when that happens.
+      firstSelectedGraphemeExtent = 0;
+      lastSelectedGraphemeExtent = 0;
+    } else {
+      final String selectedGraphemes = widget.renderObject.selection!.textInside(text);
+      firstSelectedGraphemeExtent = selectedGraphemes.characters.first.length;
+      lastSelectedGraphemeExtent = selectedGraphemes.characters.last.length;
+    }
+
+    final Rect? startHandleRect = widget.renderObject.getRectForComposingRange(TextRange(start: widget.selection.start, end: widget.selection.start + firstSelectedGraphemeExtent));
+    final Rect? endHandleRect = widget.renderObject.getRectForComposingRange(TextRange(start: widget.selection.end - lastSelectedGraphemeExtent, end: widget.selection.end));
+
+    if (firstSelectedGraphemeExtent > 0 && widget.selection.isValid && !widget.selection.isCollapsed)
+      assert(startHandleRect != null);
```

Comment:
nit: move the expression in the if statement into the assert.

Reply:
I think we only want to assert the startHandleRect and endHandleRect is not null when our graphemeExtent is greater than 0 and we confirm our current selection is valid. 

If we assert startHandleRect and endHandleRect even when our selection was invalid then the entire assertion will fail. If the selection was invalid, it is expected start/end handle rects to be null.

```dart
assert(firstSelectedGraphemeExtent > 0 && widget.selection.isValid && !widget.selection.isCollapsed && startHandleRect != null);
assert(lastSelectedGraphemeExtent > 0 && widget.selection.isValid && !widget.selection.isCollapsed && endHandleRect != null);
```

I tried something along the lines above, and received a lot of test failures.

Reply:
```assert(firstSelectedGraphemeExtent > 0 && widget.selection.isValid && !widget.selection.isCollapsed && startHandleRect != null);``` is not identical to the original logic with the `if`. It should be more of a `||` than `&&` right?

Reply:
```dart
assert(firstSelectedGraphemeExtent > 0 || widget.selection.isValid || !widget.selection.isCollapsed || startHandleRect != null);
assert(lastSelectedGraphemeExtent > 0 || widget.selection.isValid || !widget.selection.isCollapsed || endHandleRect != null);
```

Hmm did you mean something like this then? Though I admit I don't really follow this logic, can you elaborate?

The original logic said "if our selectedGraphemeExtent is greater than 0 AND our selection is valid AND the selection is not collapsed THEN assert that handle is not null"

so in our new logic should we "assert that our selectedGraphemeExtent is 0, the selection is not valid, and the selection is collapsed OR handle is not null"?